src/java.base/share/classes/jdk/internal/misc/CDS.java

```diff
@@ -59,4 +78,125 @@
+        }
+        // first part
+        if (!isBasicTypeChar(typeParts[0].charAt(0))) {
+            return false;
+        }
+        for (int i = 1; i < typeParts[0].length(); i++) {
+            char c = typeParts[0].charAt(i);
+            if (!isBasicTypeChar(c)) {
+                if (!(c >= '0' && c <= '9')) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    // return null for invalid input
+    private static Stream<String>  validateInputLines(String[] lines) {
+        ArrayList<String> list = new ArrayList<String>(lines.length);
+        for (String s: lines) {
+            String line = s.trim();
+            if (!line.startsWith("[LF_RESOLVE]") && !line.startsWith("[SPECIES_RESOLVE]")) {
+                System.out.println("Wrong prefix: " + line);
+                return null;
+            }
+
+            String[] parts = line.split(" ");
+            boolean isLF = line.startsWith("[LF_RESOLVE]");
+
+            if (isLF) {
+                if (parts.length != 4) {
+                    System.out.println("Incorrecct number of items in the line: " + parts.length);
+                    System.out.println("line: " + line);
+                    return null;
+                }
+                if (!isValidHolderName(parts[1])) {
+                    System.out.println("Invalid holder class name: " + parts[1]);
+                    return null;
+                }
+                if (!isValidMethodType(parts[3])) {
+                    System.out.println("Invalid method type: " + parts[3]);
+                    return null;
+                }
+            } else {
+                if (parts.length != 2) {
+                   System.out.println("Incorrect number of items in the line: " + parts.length);
+                   return null;
+                }
+           }
+           list.add(line);
+      }
+      return list.stream();
+    }
+
+    /**
+     * called from vm to generate MethodHandle holder classes
+     * @return @code { Object[] } if holder classes can be generated.
+     * @param lines in format of LF_RESOLVE or SPECIES_RESOLVE output
+     */
+    private static Object[] generateLambdaFormHolderClasses(String[] lines) {
+        Objects.requireNonNull(lines);
+        try {
+            Stream<String> lineStream = validateInputLines(lines);
+            if (lineStream == null) {
+                return null;
+            }
+            Map<String, byte[]> result = SharedSecrets.getJavaLangInvokeAccess().generateHolderClasses(lineStream);
+            int size = result.size();
+            Object[] retArray = new Object[size * 2];
+            int index = 0;
+            for (Map.Entry<String, byte[]> entry : result.entrySet()) {
+                retArray[index++] = entry.getKey();
+                retArray[index++] = entry.getValue();
+            };
+            return retArray;
+        } catch (Exception e) {
+            e.printStackTrace();
```

Comment:
Is this a debugging statement?   If CDS swallows the exception thrown, I think VM should emit the warning message and print the stack trace if appropriate.

Reply:
I want to print this in java since in VM it is a little complex than java. It just used to give the stack trace where the exception happened for debug purpose. Maybe I should put comment for it.